minimint-api/src/db/mod.rs

```diff
@@ -45,7 +45,41 @@ pub trait Database: Send + Sync {
 
     fn raw_find_by_prefix(&self, key_prefix: &[u8]) -> PrefixIter<'_>;
 
-    fn raw_apply_batch(&self, batch: DbBatch) -> Result<()>;
+    fn transaction<'a>(&'a self) -> Result<Box<dyn Transaction<'a>>>;
+}
+
+pub trait Transaction<'a>: 'a {
+    fn raw_insert_entry(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>>;
+
+    fn raw_insert_new_entry(&mut self, key: &[u8], value: Vec<u8>) -> Result<()> {
+        if self.raw_insert_entry(key, value)?.is_some() {
+            error!("Database replaced element! This should not happen!");
```

Comment:
Why not just one logging statement? BTW. `log` and `tracing` have support for structured logging, so a `key` can be just an associated value. https://docs.rs/tracing/latest/tracing/#recording-fields

Reply:
If this was not supposed to happen ... is it really safe to continue, instead of crashing?